internal/engine/live_update_build_and_deployer.go

```diff
@@ -47,86 +47,108 @@ func NewLiveUpdateBuildAndDeployer(dcu *containerupdate.DockerContainerUpdater,
+	containerUpdater := lubad.containerUpdaterForSpecs(specs)
+	liveUpdInfos := make([]liveUpdInfo, len(liveUpdateStateSet))
+
+	if len(liveUpdateStateSet) == 0 {
+		return nil, RedirectToNextBuilderInfof("no targets for LiveUpdate found")
 	}
 
-	liveUpdateState := liveUpdateStateSet[0]
-	iTarget := liveUpdateState.iTarget
-	state := liveUpdateState.iTargetState
-	filesChanged := liveUpdateState.filesChanged
+	for i, liveUpdateState := range liveUpdateStateSet {
+		iTarget := liveUpdateState.iTarget
+		state := liveUpdateState.iTargetState
+		filesChanged := liveUpdateState.filesChanged
 
-	span, ctx := opentracing.StartSpanFromContext(ctx, "LiveUpdateBuildAndDeployer-BuildAndDeploy")
-	span.SetTag("target", iTarget.ConfigurationRef.String())
-	defer span.Finish()
+		var changedFiles []build.PathMapping
+		var runs []model.Run
+		var hotReload bool
 
-	startTime := time.Now()
-	defer func() {
-		analytics.Get(ctx).Timer("build.container", time.Since(startTime), nil)
-	}()
+		if fbInfo := iTarget.AnyFastBuildInfo(); !fbInfo.Empty() {
+			changedFiles, err = build.FilesToPathMappings(filesChanged, fbInfo.Syncs)
+			if err != nil {
+				return store.BuildResultSet{}, err
+			}
+			runs = fbInfo.Runs
+			hotReload = fbInfo.HotReload
+		}
+		if luInfo := iTarget.AnyLiveUpdateInfo(); !luInfo.Empty() {
+			changedFiles, err = build.FilesToPathMappings(filesChanged, luInfo.SyncSteps())
+			if err != nil {
+				if pmErr, ok := err.(*build.PathMappingErr); ok {
+					// expected error for this builder. One of more files don't match sync's;
+					// i.e. they're within the docker context but not within a sync; do a full image build.
+					return nil, RedirectToNextBuilderInfof(
+						"at least one file (%s) doesn't match a LiveUpdate sync, so performing a full build", pmErr.File)
+				}
+				return store.BuildResultSet{}, err
+			}
 
-	// LiveUpdateBuildAndDeployer doesn't support initial build
-	// TODO(maia): put this check in extractImageTargetsForLiveUpdates()
-	if state.IsEmpty() {
-		return store.BuildResultSet{}, SilentRedirectToNextBuilderf("prev. build state is empty; LiveUpdate does not support initial deploy")
-	}
+			// If any changed files match a FallBackOn file, fall back to next BuildAndDeployer
+			anyMatch, file, err := luInfo.FallBackOnFiles().AnyMatch(build.PathMappingsToLocalPaths(changedFiles))
+			if err != nil {
+				return nil, err
+			}
+			if anyMatch {
+				return store.BuildResultSet{}, RedirectToNextBuilderInfof(
+					"detected change to fall_back_on file '%s'", file)
+			}
 
-	containerUpdater := lubad.containerUpdaterForSpecs(specs)
-	var changedFiles []build.PathMapping
-	var runs []model.Run
-	var hotReload bool
+			runs = luInfo.RunSteps()
+			hotReload = !luInfo.ShouldRestart()
+		}
 
-	if fbInfo := iTarget.AnyFastBuildInfo(); !fbInfo.Empty() {
-		changedFiles, err = build.FilesToPathMappings(filesChanged, fbInfo.Syncs)
-		if err != nil {
-			return store.BuildResultSet{}, err
+		liveUpdInfos[i] = liveUpdInfo{
```

Comment:
what does code this do in the case where one image has live update info and the other doesn't?

Reply:
that should never happen, `extractImageTargetsForLiveUpdates` [should only return targets with LiveUpdate or FastBuild](https://github.com/windmilleng/tilt/blob/d92acf8542722d6361421d19dacd58c51e7859b8/internal/engine/extractors.go#L63). I suppose I can put an extra check here tho